File: 2020/06_custom_customs/aoc06.jl
# Custom Customs
#
# Advent of Code 2020, day 6
# Solution by Geir Arne Hjelle, 2020-12-06

module AOC06

using Pipe


# Count number of questions to which at least one person answered yes
function count_any(group)
    @pipe group |> split .|> Set |> foldl(union, _) |> length
end


# Count number of questions to which all people answered yes
function count_all(group)
    @pipe group |> split .|> Set |> foldl(intersect, _) |> length
end


# Solve the problem for one file
function solve(input)
    # Parse input
    groups = @pipe input |> split(_, "\n\n")

    # Part 1
    part_1 = groups .|> count_any |> sum

    # Part 2
    part_2 = groups .|> count_all |> sum

    part_1, part_2
end


# Solve the problem for one file
function solve_file(file_path)
    println("\n$(file_path)")
    input = open(file_path) do fid
        read(fid, String) |> strip
    end
    input .|> solve
end

# Solve the problem for each file
[a for a in ARGS if a[1] != '-'] .|> solve_file .|> s -> join(s, "\n") |> println

end  # module

File: 2020/06_custom_customs/aoc06.py
"""Custom Customs

Advent of Code 2020, day 6
Solution by Geir Arne Hjelle, 2020-12-06
"""
# Standard library imports
import pathlib
import sys

debug = print if "--debug" in sys.argv else lambda *_: None


def count_any(group):
    """Count number of questions anyone in the group answered yes to"""
    return len(set.union(*[set(p) for p in group.split()]))


def count_all(group):
    """Count number of questions everyone in the group answered yes to"""
    return len(set.intersection(*[set(p) for p in group.split()]))


def main(args):
    """Solve the problem for all input files"""
    for file_path in (pathlib.Path(p) for p in args if not p.startswith("-")):
        solve(file_path)


def solve(file_path):
    """Solve the problem for one input file"""
    print(f"\n{file_path}:")
    groups = file_path.read_text().split("\n\n")

    # Part 1
    num_yes_any = sum(count_any(g) for g in groups)
    print(f"Anyone answered yes: {num_yes_any}")

    # Part 2
    num_yes_all = sum(count_all(g) for g in groups)
    print(f"Everyone answered yes: {num_yes_all}")


if __name__ == "__main__":
    main(sys.argv[1:])
